### ISSUE TITLE
Default value assignability in methods not checked anymore?

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#10120]
Code:

```py
from __future__ import annotations

class Box[T]:
    def __init__(self, value: T) -> None:
        self._value: T = value

    def value(self) -> T:
        return self._value
    
    def update(self, new_value: T = 0, /) -> Box[T]:
        return Box(new_value)

strbox = Box("hello")      # strbox: Box[str]

strbox2 = strbox.update()  # this call should not be allowed
                           # but pyright doesn't complain
```

I found that Pyright 1.1.382 started allowing this call (still allowed in 1.1.397). Pyright 1.1.381 correctly showed this error for the last line:
```
Argument of type "Literal[0]" cannot be assigned to parameter "new_value" of type "str" in function "update"
  "Literal[0]" is not assignable to "str"  (reportArgumentType)
```

**Comment by erictraut** [comment]
Thanks for the bug report. I agree this is a regression. I've bisected it to [this PR](https://github.com/microsoft/pyright/pull/9064), but I still need to investigate further to understand the root cause.

**Comment by erictraut** [comment]
The regression affects positional-only parameters. My regression tests didn't cover that combination, which explains why they didn't catch the bug.

This will be fixed in the next release, and I've augmented the regression tests to address this hole.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.398.
